Source/WebCore/css/CSSCounterStyle.cpp

```diff
@@ -0,0 +1,397 @@
+        return "decimal"_s;
+    return makeAtomString(rule.fallback());
+}
+
+static CounterStyle::Symbol translatePrefixFromStyleRule(const StyleRuleCounterStyle& rule)
+{
+    if (!rule.suffixCSSValue())
+        return ""_s;
+    return rule.prefix();
+}
+
+static CounterStyle::Symbol translateSuffixFromStyleRule(const StyleRuleCounterStyle& rule)
+{
+    // https://www.w3.org/TR/css-counter-styles-3/#counter-style-suffix
+    // ("." full stop followed by a space)
+    if (!rule.suffixCSSValue())
+        return "\2E\20"_s;
+    return rule.suffix();
+}
+
+void CounterStyle::extractDataFromSystemDescriptor(const StyleRuleCounterStyle& rule)
+{
+    auto systemValue = rule.systemCSSValue();
+    if (!systemValue || !systemValue->isPrimitiveValue())
+        return;
+
+    auto& primitiveSystemValue = downcast<CSSPrimitiveValue>(*systemValue);
+    ASSERT(primitiveSystemValue.isValueID() || primitiveSystemValue.isPair());
+    if (auto* pair = primitiveSystemValue.pairValue()) {
+        // This value must be `fixed` or `extends`, both of which can or must have an additional component.
+        auto secondValue = pair->second();
+        if (isExtendsSystem()) {
+            ASSERT(secondValue && secondValue->isCustomIdent());
+            m_extendsName = secondValue && secondValue->isCustomIdent() ? makeAtomString(secondValue->cssText()) : makeAtomString("decimal"_s);
+        } else if (system() == CounterStyleSystem::Fixed) {
+            ASSERT(secondValue && secondValue->isInteger());
+            // If no value is provided after `fixed`, the first synbol value is implicitly 1 (https://www.w3.org/TR/css-counter-styles-3/#first-symbol-value).
+            m_fixedSystemFirstSymbolValue = secondValue && secondValue->isInteger() ? secondValue->intValue() : 1;
+        }
+    }
+}
+
+CounterStyle::CounterStyle(Name name, CounterStyleSystem system, NegativeSymbols negativeSymbols, Symbol prefix, Symbol suffix, Ranges&& ranges, Pad pad, Name fallback, Vector<Symbol>&& symbols, CounterStyleSpeakAs speakAs, Name speakAsReference)
+: m_predefinedCounterStyle { true }, m_name { name }, m_system { system }, m_negativeSymbols { negativeSymbols }, m_prefix { prefix }, m_suffix { suffix }, m_ranges { WTFMove(ranges) }, m_pad { pad }, m_fallbackName { fallback }, m_symbols { WTFMove(symbols) }, m_speakAs { speakAs }, m_speakAsNameReference { speakAsReference }
+{ }
+
+CounterStyle::CounterStyle(const StyleRuleCounterStyle& rule)
+: m_name { rule.name() }, m_system { toCounterStyleSystemEnum(rule.systemCSSValue()) }, m_negativeSymbols { translateNegativeSymbolsFromStyleRule(rule) }, m_prefix { translatePrefixFromStyleRule(rule) }, m_suffix { translateSuffixFromStyleRule(rule) }, m_ranges { translateRangeFromStyleRule(rule) }, m_pad { translatePadFromStyleRule(rule) }, m_fallbackName { translateFallbackNameFromStyleRule(rule) }, m_symbols { translateSymbolsFromStyleRule(rule) }, m_speakAs { CounterStyleSpeakAs::Auto }, m_speakAsNameReference { ""_s }
```

Comment:
That's a lot of constructor parameters. Maybe this type could be a struct (with public fields) instead of a class? Then you just initialize it with { ... } without writing an explicit constructor.

Reply:
CounterStyle is a derived class so it won't be classified as an aggregate, right?
https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers

Reply:
The formatting does not follow WebKit style. Each field initialization should be in a separate line, indented.